### PR TITLE
Implement backend throttle

### PR DESCRIPTION
**Pull request by MiroslavGatsanoga**
**Description**:

**Related issue(s)**:

Fixes https://github.com/hashgraph/hedera-services/issues/7962

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/congestion/MonoMultiplierSources.java
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/congestion/MonoMultiplierSources.java:31]: I think we should introduce `@GasPriceMultiplier` and `@GenericPriceMultiplier` as well as bindings in `FeesInjectionModule` exactly like in mono:
```
@Binds
@Singleton
@GenericPriceMultiplier
FeeMultiplierSource bindGenericFeeMultiplierSource(TxnRateFeeMultiplierSource txnRateFeeMultiplierSource);

@Binds
@Singleton
@GasPriceMultiplier
FeeMultiplierSource bindGasFeeMultiplierSource(GasPriceMultiplierSource gasPriceMultiplierSource);

```

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/NetworkUtilizationManager.java
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/NetworkUtilizationManager.java:24]: Can we add a similar comment as `NetworkUtilization`? It seems useful to understand what exactly this interface is for.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java:49]: nit: should this be called `multiplierSources`?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java:426]: Is this still to be implemented or did you just forget about that commented code?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java:41]: Why do we need this extra parameter?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/congestion/MonoMultiplierSources.java
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/congestion/MonoMultiplierSources.java:31]: Oh, you have the same thing but without the annotations inside `ThrottleInjectionModule` 👍

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:344]: Can't we just move that before we charge the user(line 336) so we don't charge the user in the first place? Or do that so we can have an event history of some sort?

**Comment by vtronkov** [comment]
Good job. Other than small comments, LGTM. Should we also update the `throttles.md` file? Maybe we should first wait to have approvals and at the end add details there.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java:426]: Good point, I missed to add a TODO here because I was not sure if we need to do query throttling in the backend throttle (I see there is an implementation in mono for this but did not find any actual usages).

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:344]: I implemented it that way because that's how it's done in mono.
If I move it up I would need to have a conditional and for this case we will need to charge the node and network components only and otherwise charge all three components i.e.
```
try {
       if (!isPayerThrottleExempt && networkUtilizationManager.wasLastTxnGasThrottled()) {
              final var nodeAndNetworkFees = new Fees(fees.nodeFee(), fees.networkFee(), 0L);
              feeAccumulator.charge(payer, nodeAndNetworkFees);
              throw new HandleException(CONSENSUS_GAS_EXHAUSTED);
       } else {
              feeAccumulator.charge(payer, fees);
       }
...
```
It doesn't look much different to me, not sure what we gain if I do it that way?

Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:344]: Oh, the fees are different. Gotcha. I thought that the fees were the same and that we were charging the user with a fee and then if the last transaction was gas throttled - we just refunded the same fee.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/CongestionThrottleService.java
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/CongestionThrottleService.java:32]: Should we register this service in `Hedera`'s class constructor?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:343]: Should we also set the `payer` here? The idea is that currently it might work as expected, but someone can write a code that sets `payer` to something else in the future - we don't have control over it. If we set the `fees` and `payer` and then throw the exception we are sure that we are going to charge the correct user the correct amount.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java:448]: Just FYI - I implemented this method `public boolean shouldThrottleQuery(HederaFunctionality queryFunction, Instant now, Query query)` in the FE throttle branch

**Comment by vtronkov** [comment]
> Good job. Other than small comments, LGTM. Should we also update the `throttles.md` file? Maybe we should first wait to have approvals and at the end add details there.

I was looking at the Backlog and I saw that we have a story for adding the throttle documentation https://github.com/hashgraph/hedera-services/issues/7441. You can leave it for now if you want.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java:471]: It would be best not to concatenate a String when calling the constructor of `StringBuilder`. What is likely to happen is that the JVM will create a first StringBuilder to concatenate the String, and then the result is used to create a second StringBuilder. (At least, this happened the last time I analyzed such code a couple of years ago.)

We can check if the compiler became smarter and can identify this pattern. Or we can fix this manually:
```
var sb = new StringBuilder("Resolved handle throttles (after splitting capacity ").append(capacitySplit).append(" ways) - \n");
```

I know it is ugly, but the outcome is better. 🙂

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java:450]: I'd always try to reduce the scope of a local variable to where it is used. For example, here, `capacity` is not used (or even initialized) until this line. You could rewrite it like this:
```
if (contractsConfig.throttleThrottleByGas() && contractsConfig.maxGasPerSec() == 0) {
    log.warn("Consensus gas throttling enabled, but limited to 0 gas/sec");
    return;
}
final long capacity = contractsConfig.maxGasPerSec();
```

It's not mandatory. The current code works just fine.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java:456]: Might be easier to read with `String.format()` (or using the formatting-feature in `log.info()` directly).

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/NetworkUtilizationManager.java
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/NetworkUtilizationManager.java:29]: Can you please add JavaDocs for all methods?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleInjectionModule.java
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleInjectionModule.java:27]: Is this used anywhere?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java:55]: Looks like Spotless screwed up the format of your comment.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java:80]: Does it make sense to inline `track()`? `trackFeePayments()` can call this method directly.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java:176]: Does it make sense to move these methods to `PbjConverter`?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:372]: Can you please create a ticket to track this?

File: hedera-node/settings.gradle.kts
Comment (MiroslavGatsanoga) [reply hedera-node/settings.gradle.kts:92]: Yes, https://github.com/hashgraph/hedera-protobufs/pull/304 looks ready to be merged, I don't have permissions to merge it though. Once it's merged I will update this.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/congestion/MonoMultiplierSources.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/congestion/MonoMultiplierSources.java:44]: Added comment.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java]: Such method was introduced in different PR, updated the code to use it.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java:91]: Yes, for handle throttle the passed value should be `consensusTime`, renamed it.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java:156]: It was in mono that way and did not actually think about why. It looks like it should not matter, updated the code to assign first and then check.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java:471]: Fixed.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java:450]: Updated.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java:456]: Fixed, using `log.info()` formatting.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleInjectionModule.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleInjectionModule.java:27]: I had some temporary implementation at some point using it but currently it's not used, will remove it.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java:80]: Updated.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java:176]: Updated.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:385]: Created https://github.com/hashgraph/hedera-services/issues/8874.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java]: Updated.

File: hedera-node/settings.gradle.kts
Comment (netopyr) [reply hedera-node/settings.gradle.kts:92]: I just merged it

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HandleThrottleAccumulator.java:81]: Added.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/NetworkUtilizationManager.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/NetworkUtilizationManager.java]: Added.

File: hedera-node/settings.gradle.kts
Comment (MiroslavGatsanoga) [reply hedera-node/settings.gradle.kts]: Fixed.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/CongestionThrottleService.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/CongestionThrottleService.java:56]: Not sure, do we need it for the first transaction handle in which we try to read the throttle usage from state i.e. the logic in `NetworkUtilizationManagerImpl.resetFrom`?

**Review by netopyr — CHANGES_REQUESTED** [review]
The code itself looks good, but the HAPI tests do not start. Maybe this is caused by the missing data initialization.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java:169]: Can we be sure that all code can deal with `null` in the `List`? Should we filter `null`s?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java:172]: Same here

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java
Comment (netopyr) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java:1554]: We have `HapiUtils.asInstant()` to convert a `Timestamp` to an `Instant`.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java
Comment (netopyr) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java:1564]: We have `HapiUtils.asTimestamp()` to convert an `Instant` to a `Timestamp`.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java:169]: Yes, actually we can just filter the `null`s, will update that.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java]: Updated.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/pbj/PbjConverter.java]: Updated.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java]: Updated.

**Review by netopyr — APPROVED** [review]
LGTM

**Pull request merged**
